### PR TITLE
feat: add peering config to mesh CRD

### PR DESCRIPTION
**Pull request by DanStough**
Changes proposed in this PR:
- Adding a new field to the `mesh` CRD that will be used to turn on routing of peering control-plane traffic through a mesh gateway(s). [Consul PR link](https://github.com/hashicorp/consul/pull/14475)

How I've tested this PR:
* I've just been running unit tests. Right now nothing in Consul is listening to this flag - more integration testing required at that point. Also assuming that the changelog will be deferred until that point (please correct me if this isn't convention).

How I expect reviewers to test this PR:
* Hopefully CI tests are sufficient.

Checklist:
- [X] Tests added
- [ ] CHANGELOG entry added (TBD)
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: control-plane/api/v1alpha1/mesh_webhook.go [hunk 1]
```diff
@@ -59,7 +63,19 @@ func (v *MeshWebhook) Handle(ctx context.Context, req admission.Request) admissi
 		}
 	}
 
-	return admission.Allowed(fmt.Sprintf("valid %s request", mesh.KubeKind()))
+	return common.ValidateConfigEntry(ctx, req, v.Logger, v, &mesh, v.ConsulMeta)
```
Comment (DanStough) [inline control-plane/api/v1alpha1/mesh_webhook.go:66]: Double check my work but I think this was a bug that meant we weren't running any of the existing TLS validation.

File: control-plane/api/v1alpha1/mesh_types.go [hunk 2]
```diff
@@ -194,11 +206,12 @@ func (in *Mesh) MatchesConsul(candidate capi.ConfigEntry) bool {
 	return cmp.Equal(in.ToConsul(""), configEntry, cmpopts.IgnoreFields(capi.MeshConfigEntry{}, "Partition", "Namespace", "Meta", "ModifyIndex", "CreateIndex"), cmpopts.IgnoreUnexported(), cmpopts.EquateEmpty())
 }
 
-func (in *Mesh) Validate(_ common.ConsulMeta) error {
+func (in *Mesh) Validate(consulMeta common.ConsulMeta) error {
 	var errs field.ErrorList
 	path := field.NewPath("spec")
 
 	errs = append(errs, in.Spec.TLS.validate(path.Child("tls"))...)
+	errs = append(errs, in.Spec.Peering.validate(path.Child("peering"), consulMeta.PartitionsEnabled, consulMeta.Partition)...)
```
Comment (DanStough) [inline control-plane/api/v1alpha1/mesh_types.go:214]: There is an additional validation in Consul that mesh CRs are always in the default namespace, but that always seems to be implied here on creation. I don't see where namespace is transferred in `ToConsul`.

File: control-plane/api/v1alpha1/mesh_webhook.go (hunk 1 above)
Comment (thisisnotashwin) [reply control-plane/api/v1alpha1/mesh_webhook.go:66]: Excellent Catch!! thank you!!

**Review by thisisnotashwin — APPROVED** [review]
This looks great! Before we merge this though, is the `peerThroughMeshGateways` already supported in 1.13 or will this be a 1.14 feature? We might want to wait to merge it if it is for 1.14.

**Comment by DanStough** [comment]
@thisisnotashwin this will be for 1.14. I guess a few weeks before the release is a good time to merge?

**Comment by thisisnotashwin** [comment]
> @thisisnotashwin this will be for 1.14. I guess a few weeks before the release is a good time to merge?

Yup!! that sounds perfect!! Thanks Dan!

**Comment by david-yu** [comment]
@DanStough Should this go out for 1.14 beta? If so, I think this needs a re-base off of main

**Comment by DanStough** [comment]
@david-yu . Nope, not required for the beta. I will rebase shortly though.

**Comment by DanStough** [comment]
@ndhanushkodi would you be available to review? I finally rebased everything so there shouldn't be conflicts.

**Review by ndhanushkodi — APPROVED** [review]
This is great and thank you for catching the missing validation!

**Comment by ndhanushkodi** [comment]
Will wait for peering acceptance tests to pass then merge

**Pull request merged**